File: src/montecarlo/montecarlo_.py
import random
from time import time
from typing import Optional

from game import Game, State, Action
from utils import argmax
from . import MonteCarloScoreStrategy, MonteCarloStats, MonteCarloNodeStats, MonteCarloNode, MonteCarloBestMoveStrategy
import logging

logger = logging.getLogger(__name__)


# https://medium.com/@quasimik/implementing-monte-carlo-tree-search-in-node-js-5f07595104df


class MonteCarlo:

    def __init__(self, game: Game, score_strategy: MonteCarloScoreStrategy,
                 best_action_strategy: MonteCarloBestMoveStrategy, win_score=1, draw_score=0.5):
        self._game = game
        self._score_strategy = score_strategy
        self._best_action_strategy = best_action_strategy
        self._win_score = win_score
        self._draw_score = draw_score
        self._nodes: dict[int, MonteCarloNode] = {}
        self._root_node: MonteCarloNode = None

    def stats(self, state: State) -> MonteCarloStats:
        """Return MCTS statistics for this node and children nodes."""
        node: MonteCarloNode = self._nodes[state.history_hash()]
        stats = MonteCarloStats(node.n_simulations, node.win_score)

        for child in node.children.values():
            if child.node is None:
                stats.children.append(
                    MonteCarloNodeStats(child.action, 0, 0))
            else:
                stats.children.append(MonteCarloNodeStats(
                    child.action, child.node.n_simulations, child.node.win_score))
        return stats

    def make_node(self, state: State):
        """If state does not exist, create dangling node.
        \n@param {State} state - The state to make a dangling node for; its parent is set to null."""

        action_history = state.history_hash()
        if action_history not in self._nodes:
            unexpanded_actions = self._game.legal_actions(state).copy()
            node = MonteCarloNode(None, None, state, unexpanded_actions)
            self._nodes[action_history] = node

    def update_root_node(self, action: Action, state: State):
        action_history = state.history_hash()
        node = self._nodes.get(action_history)

        if node is None and self._root_node is not None and action in self._root_node.children:
            logger.debug(f'Node found for action {action}')
            child_node = self._root_node.children[action]
            action = child_node.action
            node = child_node.node

        if node is None:
            unexpanded_actions = self._game.legal_actions(state).copy()
            node = MonteCarloNode(self._root_node, action, state, unexpanded_actions)

        self._nodes[action_history] = node
        self._root_node = node

    def best_action(self, state: State) -> Action:
        """Get the best action from available statistics."""
        node = self._nodes.get(state.history_hash())
        if node is None:
            raise Exception('Run search before getting best move')

        # If not all children are expanded, not enough information
        if not node.is_fully_expanded():
            raise Exception("Not enough information!")

        best_action = argmax(
            lambda action: self._best_action_strategy.score(node.child_node(action)),
            node.all_actions())

        logger.debug(f'Best action = {best_action}')
        return best_action

    def run_search(self, state: State, timeout_s: int):
        """From given state, run as many simulations as possible until the time limit (in seconds), building statistics."""
        self.make_node(state)

        node = self._nodes[state.history_hash()]
        unexpanded = len(node.unexpanded_actions())
        total = len(node.all_actions())
        logger.debug(f'Unexpanded = {unexpanded} / {total}')

        end = time() + timeout_s
        while time() < end:
            node = self._select(state)
            winner = self._game.winner(node.state)

            # if the match is not closed and there are possible actions from the selected node
            if winner is None and not node.is_leaf():
                node = self._expand(node)
                winner = self._simulate(node)

            if winner is None:
                raise Exception('No actions available')

            self._backpropagate(node, winner)

        node = self._nodes[state.history_hash()]
        unexpanded = len(node.unexpanded_actions())
        total = len(node.all_actions())
        logger.debug(f'Unexpanded = {unexpanded} / {total}')

    def _select(self, state: State):
        """Phase 1, Selection: Select until not fully expanded OR leaf."""
        node = self._nodes[state.history_hash()]

        while node.is_fully_expanded() and not node.is_leaf():
            best_action = argmax(
                lambda action: node.child_node(action).score(self._score_strategy),
                node.all_actions())

            node = node.child_node(best_action)

        return node

    def _expand(self, node: MonteCarloNode):
        """Phase 2, Expansion: Expand a random unexpanded child node."""
        unexpanded_actions = node.unexpanded_actions()
        if len(unexpanded_actions) == 0:
            logger.critical(f'Len = 0, state = {node.state}')
        action = random.choice(unexpanded_actions)

        child_state = self._game.next_state(node.state, action)
        child_unexpanded_actions = self._game.legal_actions(child_state)
        child_node = node.expand(action, child_state, child_unexpanded_actions)

        logger.debug(f'Expanding action {action}, turn = {child_state.turn}')

        return child_node

    def _simulate(self, node: MonteCarloNode, choice=random.choice) -> Optional[int]:
        """Phase 3, Simulation: Play game to terminal state using random actions, return winner."""
        state = node.state
        winner = self._game.winner(state)

        while winner is None:
            action = choice(self._game.legal_actions(state))
            state = self._game.next_state(state, action)
            winner = self._game.winner(state)

        return winner

    def _backpropagate(self, node: MonteCarloNode, winner: int):
        """Phase 4, Backpropagation: Update ancestor statistics."""
        while node is not None:
            node.n_simulations += 1

            if winner == Game.DRAW:
                node.win_score += self._draw_score

            # if black wins, the win counter of each visited white node is increased, as white statistics are used for black's choice (and viceversa)
            elif node.state.is_player_turn(self._game.previous_player(winner)):
                node.win_score += self._win_score
            else:
                node.win_score -= self._win_score

            node = node.parent

File: java/src/main/java/it/ai/montecarlo/MonteCarlo.java
package it.ai.montecarlo;

import it.ai.game.Action;
import it.ai.game.Coords;
import it.ai.game.Game;
import it.ai.game.State;
import it.ai.montecarlo.MonteCarloStats.MonteCarloNodeStats;
import it.ai.montecarlo.strategies.bestaction.MonteCarloBestActionStrategy;
import it.ai.montecarlo.strategies.score.MonteCarloSelectionScoreStrategy;
import it.ai.montecarlo.strategies.winscore.WinScoreStrategy;
import it.ai.util.MathUtils;
import it.ai.util.RandomUtils;
import lombok.Getter;

import java.util.*;
import java.util.function.Function;
import java.util.logging.Logger;
import java.util.stream.Collectors;

public class MonteCarlo {
    protected final Logger logger = Logger.getLogger(MonteCarlo.class.getName());

    protected final Game game;
    protected final MonteCarloSelectionScoreStrategy selectionScoreStrategy;
    protected final MonteCarloBestActionStrategy bestActionStrategy;
    protected final WinScoreStrategy winScoreStrategy;

    protected Map<State, MonteCarloNode> nodes = new HashMap<>();

    public MonteCarlo(Game game, MonteCarloSelectionScoreStrategy selectionScoreStrategy, MonteCarloBestActionStrategy bestActionStrategy, WinScoreStrategy winScoreStrategy) {
        this.game = game;
        this.selectionScoreStrategy = selectionScoreStrategy;
        this.bestActionStrategy = bestActionStrategy;
        this.winScoreStrategy = winScoreStrategy;
    }

    /***
     * If state does not exist, create dangling node.
     *
     * @param state The state to make a dangling node for; its parent is set to null.
     */
    public void makeNode(State state) {
        if (!nodes.containsKey(state)) {
            Iterable<Action> unexpandedActions = game.getValidActions(state);
            MonteCarloNode node = new MonteCarloNode(null, null, state, unexpandedActions);
            nodes.put(state, node);
        }
    }

    public void updateNodes(State state) {
        //TODO: usare la libreria java che storicizza in memory e on disk, basta fare map.clearWithExpires(). dbMap
        MonteCarloNode node = nodes.get(state);

        nodes = new HashMap<>(); //TODO: Forse meglio HashTable
        if (node != null) {
            logger.fine("Node found.");
            nodes = node.getExpandedNodes().collect(
                    Collectors.toMap(MonteCarloNode::getState, Function.identity()));
            nodes.put(state, node);
//            System.gc();
        }
    }

    /***
     *
     * @return the best action from available statistics.
     */
    public Action bestAction(State state) {
        MonteCarloNode node = nodes.get(state);
        if (node == null) throw new RuntimeException("Run search before getting best move.");

        // If not all children are expanded, not enough information
        if (!node.isFullyExpanded()) {
            long unexpanded = node.getUnexpandedActions().count();
            long total = node.getAllActions().count();
            long expanded = total - unexpanded;
            logger.severe("Expanded = " + expanded + " / " + total);
            throw new RuntimeException("Not enough information!\nExpanded = " + expanded + " / " + total);
        }

        Action bestAction = MathUtils.argmax(node.getAllActions()::iterator,
                action -> bestActionStrategy.score(node.getChildNode(action)));

        logger.fine("Best action = ".concat(bestAction.toString()));
        return bestAction;
    }

    /***
     * From given state, run as many simulations as possible until the time limit (in seconds), building statistics.
     * @param timeout_s timeout in seconds
     */
    public void runSearch(State state, int timeout_s) {
        updateNodes(state);
        makeNode(state);

        long end = System.currentTimeMillis() + timeout_s * 1000L;
        while (System.currentTimeMillis() < end) {
            MonteCarloNode node = selection(state);
            Optional<Integer> winner = game.getWinner(node.getState());

            DistanceFromFinalState distance = new DistanceFromFinalState();
            //if the match is not closed and there are possible actions from the selected node
            if (!winner.isPresent() && !node.isLeaf()) {
                node = expansion(node);
                winner = simulation(node, RandomUtils::choice);
            }

            if (!winner.isPresent()) throw new RuntimeException("No actions available.");

            backpropagation(node, winner.get());
        }

        MonteCarloNode node = nodes.get(state);
        long unexpanded = node.getUnexpandedActions().count();
        long total = node.getAllActions().count();
        long expanded = total - unexpanded;
        logger.fine("Expanded = " + expanded + " / " + total);
        logger.fine("n_simulations " + node.numberOfSimulations() + ", win_score " + node.winScore());
    }

    /***
     * Phase 1, Selection: Select until not fully expanded OR leaf.
     */
    protected MonteCarloNode selection(State state) {
        MonteCarloNode node = nodes.get(state);

        while (node.isFullyExpanded() && !node.isLeaf()) {

            MonteCarloNode finalNode = node;
            Action bestAction = MathUtils.argmax(node.getAllActions()::iterator,
                    action -> finalNode.getChildNode(action).score(selectionScoreStrategy));

            node = node.getChildNode(bestAction);
        }

        return node;
    }

    /***
     * Phase 2, Expansion: Expand a random unexpanded child node.
     */
    protected MonteCarloNode expansion(MonteCarloNode node) {
        List<Action> unexpandedActions = node.getUnexpandedActions().collect(Collectors.toList());
        if (unexpandedActions.isEmpty())
            logger.severe("Len = 0, state = ".concat(node.getState().toString()));

        Action action = RandomUtils.choice(unexpandedActions);

        State childState = game.nextState(node.getState(), action);
        Iterable<Action> childUnexpandedActions = game.getValidActions(childState);
        MonteCarloNode childNode = node.expand(action, childState, childUnexpandedActions);

        nodes.put(childState, childNode);

        //if child_state.turn == 0:
        //logger.debug(f'Expanding action {action}, parent = {self._nodes[node.state.history_hash()].action}, turn = {child_state.turn}')
        return childNode;
    }

    /***
     * Phase 3, Simulation: Play game to terminal state using random actions, return winner.
     */
    protected Optional<Integer> simulation(MonteCarloNode node, Function<List<Action>, Action> actionChoicePolicy) {
        State state = node.getState();
        Optional<Integer> winner = game.getWinner(state);

        while (!winner.isPresent()) {
            List<Action> validActions = game.getValidActions(state);
            if (validActions.isEmpty())
                throw new RuntimeException("No valid actions for state" + state);

            Action action = actionChoicePolicy.apply(validActions);
            state = game.nextState(state, action);
            winner = game.getWinner(state);
//            distance.increment();
        }

        return winner;
    }

    /***
     * Phase 4, Backpropagation: Update ancestor statistics.
     */
    protected void backpropagation(MonteCarloNode node, int winner) {
//        Map<Integer, Set<Coords>> cellsOccupiedByPlayers = new HashMap<>();
//        if (finalState != null)
//            cellsOccupiedByPlayers = getCellsOccupiedByPlayers(finalState);

        int nextPlayer = game.nextPlayer(winner);

        while (node != null) {
            node.numberOfSimulations += 1;

            incrementScore(node, winner, nextPlayer);

//            incrementRaveScore();


//            distance.increment();
            node = node.getParent();
        }
    }

    private void incrementScore(MonteCarloNode node, int winner, int nextPlayer) {
        double reward;
        if (winner == Game.DRAW)
            reward = winScoreStrategy.drawScore();

            //Score of child node is used by the parent.
            //Therefore, we increment the child node score if the parent player has won.
        else if (node.getState().isPlayerTurn(nextPlayer))
            reward = winScoreStrategy.winScore();
        else
            reward = winScoreStrategy.loseScore();

        node.winScore += reward;
    }

    private Map<Integer, Set<Coords>> getCellsOccupiedByPlayers(State state) {
        return game.getPlayers().collect(Collectors.toMap(Function.identity(),
                player -> game.cellsOccupiedBy(state, player).collect(Collectors.toSet())));
    }

    private void incrementRaveScore() {
//        if (finalState != null) {
//            Set<Coords> occupiedCells = cellsOccupiedByPlayers.get(node.getState().getTurn());
//            double finalReward = -reward;
//            node.getChildrenNodes().forEach(child_node -> {
//                //If the pawn moved is still in the final state and the player has won (lost), the action was good(bad)
//                //Therefore, we increase the child score
//
//                if (occupiedCells.contains(child_node.getAction().getTo())) {
//                    child_node.numberOfRave += 1;
//                    child_node.raveScore += finalReward;
//                }
//            });
//        }
    }

    /***
     * Return MCTS statistics for this node and children nodes.
     */
    public MonteCarloStats getStats(State state) {
        MonteCarloNode node = nodes.get(state);
        MonteCarloStats stats = new MonteCarloStats(node.numberOfSimulations, node.winScore,
                node.numberOfRave, node.raveScore);

        for (MonteCarloNode.MonteCarloChild child : node.getChildren().values()) {
            if (child.getNode() == null)
                stats.getChildren().add(new MonteCarloNodeStats(child.getAction()));
            else
                stats.getChildren().add(
                        new MonteCarloNodeStats(child.getAction(),
                                child.getNode().numberOfSimulations, child.getNode().winScore,
                                child.getNode().numberOfRave, child.getNode().raveScore));
        }
        return stats;
    }

    /***
     * Commentare tutta sta porcheria RAVE che tanto non funziona (per ora)
     * TODO: add move priority in order to choice move in simulation phase
     * They can be summarised by applying four prioritised rules after any opponent move a:
     *      1. If a put some of our stones into atari, play a saving move at random.
     *      2. Otherwise, if one of the 8 intersections surrounding a matches a simple pattern for cutting or hane,
     *          randomly play one.
     *      3. Otherwise, if any opponent stone can be captured, play a capturing move at random.
     *      4. Otherwise play a random move.
     *
     *  Successivamente è possibile usare supervised learning per stimare i pesi della default_policy
     */

    protected static class DistanceFromFinalState {
        @Getter
        private int value = 1;

        public void increment() {
            value++;
        }
    }

}
